### PR TITLE
Mark constructor of `BinaryHeap` as const fn

### LINKED ISSUE #112353: Constification of BinaryHeap construction
Is there any reason this isn't const yet? This should only depend on vec's construction being const which it is.

### PR DESCRIPTION
**Pull request by Coekjan**
#112353

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]
Thank you!

File: library/alloc/src/collections/binary_heap/mod.rs [hunk 1]
```diff
@@ -434,8 +434,9 @@ impl<T: Ord> BinaryHeap<T> {
     /// heap.push(4);
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
+    #[rustc_const_unstable(feature = "const_binary_heap_constructor", issue = "112353")]
     #[must_use]
-    pub fn new() -> BinaryHeap<T> {
+    pub const fn new() -> BinaryHeap<T> {
         BinaryHeap { data: vec![] }
```
Comment (oli-obk) [inline library/alloc/src/collections/binary_heap/mod.rs:440]: if you use `Vec::new()` here, then `BinaryHeap::new` can become const stable. `vec![]` goes through a code path that assumes it needs to allocate.

Comment (Coekjan) [reply library/alloc/src/collections/binary_heap/mod.rs:440]: Perhaps I should have marked onstification of `new` and `new_in` as different features, which may helps to stablize const `new` function.

Comment (Coekjan) [reply library/alloc/src/collections/binary_heap/mod.rs:440]: https://github.com/rust-lang/rust/blob/bbcc1691a4646a6d2ebb2189bdff5b878407fbe6/library/alloc/src/macros.rs#L42-L57

It seems that the macro `vec![]` (nothing filled within the bracket) just expands into `Vec::new()` **without allocator**.

Thus, constification of `BinaryHeap::new` is able to be stablized, if the `new` function does not share the mark `rust_const_unstable(feature = "const_binary_heap_constructor")` with `BinaryHeap::new_in`.

**Pull request merged**